### PR TITLE
Adding unbiased superclusters to miniAOD and EGM nano

### PR DESCRIPTION
**Pull request by SanghyunKo**
#### PR description:

This PR addresses two issues:

1. Replace the ratio between corrected & uncorrected energy in Run 2 NanoAOD with the uncorrected energy itself (to avoid issues like #46046 - if the correction is 0, we cannot recover the energy before applying the correction using the ratio). The NanoDQM plots are modified accordingly.

2. Currently, EGM still relies on AOD to derive central reco SFs. This is because the supercluster (SC) collection stored in MiniAOD is highly biased since we only store the SC that already reconstructed electron & photon objects. This PR attempts to improve the situation by migrating from the legacy AOD-based workflow to the (EGM) nanoAOD-based workflow. A minimal change in miniAOD is required to remove the biases implied in the current miniAOD. Furthermore, the SC track isolation suggested in #42007 is also incorporated into the EGM NanoAOD.

#### PR validation:

The expected increase in miniAOD file size is marginal - DY: 0.27%, TTto2L2Nu: 0.62%, TTto4Q: 1% (tested 1k events with SC pt > 5 GeV). The increase in EGM nanoAOD due to the new SC collection is approximately 2%. A quick validation on the new track isolation variable has been performed with the DY & TTto4Q samples:

DY
scTrkIso_DY

TTto4Q
scTrkIso_TTto4Q

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

No backport is needed (this PR targets Nano v15).

### CODE REVIEW

**Comment by hqucms** [comment]
enable nano

**Comment by hqucms** [comment]
please test

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 1]
```diff
@@ -433,7 +433,7 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
         electronTable.variables,
         pt = Var("pt*userFloat('ecalTrkEnergyPostCorrNew')/userFloat('ecalTrkEnergyPreCorrNew')", float, precision=-1, doc="p_{T}"),
         energyErr = Var("userFloat('ecalTrkEnergyErrPostCorrNew')", float, precision=6, doc="energy error of the cluster-track combination"),
-        eCorr = Var("userFloat('ecalTrkEnergyPostCorrNew')/userFloat('ecalTrkEnergyPreCorrNew')", float, doc="ratio of the calibrated energy/miniaod energy"),
+        ptPreCorr = Var("pt", float, doc="pt of the electron before energy corrections"),
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:436]: Please update the nanoDQM with this change.

File: PhysicsTools/NanoAOD/python/photons_cff.py [hunk 2]
```diff
@@ -283,7 +283,7 @@ def make_bitmapVID_docstring(id_modules_working_points_pset):
     photonTable.variables,
     pt = Var("pt*userFloat('ecalEnergyPostCorrNew')/userFloat('ecalEnergyPreCorrNew')", float, precision=-1, doc="p_{T}"),
     energyErr = Var("userFloat('ecalEnergyErrPostCorrNew')",float,doc="energy error of the cluster from regression",precision=6),
-    eCorr = Var("userFloat('ecalEnergyPostCorrNew')/userFloat('ecalEnergyPreCorrNew')",float,doc="ratio of the calibrated energy/miniaod energy"),
+    ptPreCorr = Var("pt",float,doc="pt of the photon before energy corrections"),
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/photons_cff.py:286]: Please update the nanoDQM with this change.

**Comment by jfernan2** [comment]
please test

**Comment by stahlleiton** [comment]
@mandrenguyen : I think it would also be good if we could backport this to 14_1_X, at least to be able to use miniAOD to derive the egamma reco SFs for PbPb data.

**Comment by SanghyunKo** [comment]
Two updates:
1. Reflect the variable name change from `eCorr` to `ptPreCorr` in the nano DQM (my bad!).
2. From the cms-bot, I realized that there was a PR #42007 that attempted to address the very same issue, but in the miniAOD level instead of nano. I incorporated one of the proposed changes there - track isolation for the SC. Because I didn't create the PAT version of SCs, the track isolation values are calculated on the fly from miniAOD, analogous to the HEEP track isolation (nevertheless, it should do a fairly similar task). I ran a quick validation by calculating SC track isolation using 1k events of DY & TTto4Q samples.

DY
scTrkIso_DY

TTto4Q
scTrkIso_TTto4Q

**Comment by hqucms** [comment]
please test

**Comment by SanghyunKo** [comment]
We just had a comment to accommodate a possible extension of the official reco SF coverage to the electrons with lower Pt (apology for the frequent commits... hopefully this is the last change from the EGM side)

We've checked the increase in the miniAOD file size with the new threshold - DY: 0.27%, TTto2L2Nu: 0.62%, TTto4Q: 1%.

**Comment by hqucms** [comment]
please test

**Comment by hqucms** [comment]
+1

**Comment by hqucms** [comment]
@SanghyunKo Could you please update the PR description with the changes you made later, i.e. https://github.com/cms-sw/cmssw/pull/46423#issuecomment-2424353787 and https://github.com/cms-sw/cmssw/pull/46423#issuecomment-2426745770?

**Comment by SanghyunKo** [comment]
@hqucms no problem, done.

**Comment by hqucms** [comment]
> @hqucms no problem, done.

Thanks a lot!

**Comment by jfernan2** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Comment by RSalvatico** [comment]
type egamma

**Pull request merged**